File: go/SwapNodesInPairs/SwapNodesInPairs.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time -> 0(N) Space -> 0(1)
func swapPairs(head *ListNode) *ListNode {
	swap := head
	if head != nil && head.Next != nil {
		swap = head.Next
	}
	var prev *ListNode
	for head != nil && head.Next != nil {
		next := head.Next.Next
		tmpHead := head
		head = head.Next
		if prev != nil {
			prev.Next = head
		}
		head.Next = tmpHead
		head.Next.Next = next
		prev = head.Next
		head = next
	}
	return swap
}

func main() {
	g := &ListNode{Val: 5}
	f := &ListNode{Val: 4, Next: g}
	d := &ListNode{Val: 3, Next: f}
	s := &ListNode{Val: 2, Next: d}
	a := &ListNode{Val: 1, Next: s}

	head := swapPairs(a)

	for head != nil {
		fmt.Println(head.Val)
		head = head.Next
	}
}

File: python/swapNodesInPairs.py
class ListNode:
    def __init__(self, val=0, next=None) -> None:
        self.val = val
        self.next = next


# Time -> 0(N) Space -> 0(1)
def swapPairs(head) -> ListNode:
    swap = head
    prev = None
    if head is not None and head.next is not None:
        swap = head.next

    while head is not None and head.next is not None:
        next = head.next.next
        tmpHead = head
        head = head.next
        if prev is not None:
            prev.next = head
        head.next = tmpHead
        head.next.next = next
        prev = head.next
        head = next

    return swap
